### PR TITLE
Release 0.7.8 — reconcile stale spec frontmatter status (no more false defers)

### PR DESCRIPTION
**Pull request by pbean**
## Summary

Fixes completed `bmad-dev-auto` work being **falsely deferred** when the skill leaves the spec
frontmatter `status` behind its prose.

The decoupled `bmad-dev-auto` skill can finalize a run in its `## Auto Run Result` prose
(`Status: done`) yet leave the YAML frontmatter at the template default `draft`. Every orchestrator
gate reads frontmatter, so `_post_dev_state_sync` no-op'd (sprint/ledger never advanced) and the
story or sweep bundle deferred — and with `rollback_on_failure` on, the tested work was reverted and
lost. (Observed in the wild: a fully-implemented, 1442-test-green deferred-work bundle deferred with
`spec status is 'draft', expected 'done'`.)

## Fix (Approach: reconcile, not escalate)

`engine._reconcile_generic_terminal_status` runs in the shared `Engine._dev_phase` immediately
before `_post_dev_state_sync`, so it covers **both** the story (sprint) and sweep-bundle (ledger)
paths. When the terminal `## Auto Run Result` Status is `done` and the frontmatter is at a
non-terminal status (`devcontract.RECONCILABLE_FROM`: `""`/`draft`/`ready-for-dev`/`in-progress`),
it advances the frontmatter to the success status via `devcontract.reset_spec_status` and journals
`spec-status-reconciled`.

**This does not violate the "never trust prose for a gate" doctrine.** Reconcile repairs a
bookkeeping field only; every deterministic gate still runs afterward against real on-disk/git state:

- `has_changes_since(baseline)` — real worktree diff
- dw-id cross-check (bundles)
- the verify-command suite
- the on-disk ledger check (review-bundle)

So a spec that lied (no diff / wrong dw-ids / broken build) still fails its gate. `blocked` is never
reconciled (guard: prose must be `done`); an already-terminal or unknown frontmatter status is never
overridden (idempotent, never-regress).

The companion upstream skill fix is bmad-code-org/BMAD-METHOD#2536 (make the Finalize step write
frontmatter `status` explicitly); this orchestrator change is the durable backstop regardless of
skill version.

## Tests

- `test_devcontract.py` — `RECONCILABLE_FROM` guard + reset-from-each-reconcilable-value
- `test_engine.py` — advance / idempotent-when-already-done / skip-blocked-prose / does-not-bypass-the-no-change-gate (story path)
- `test_sweep.py` — the exact DW-bundle regression: stale-frontmatter bundle now closes the ledger

Full suite **1194 green**, `trunk check` clean. Version bumped `0.7.7 → 0.7.8` (sync_version), CHANGELOG updated.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed completed work sometimes staying deferred when spec frontmatter status was stale, by reconciling frontmatter status from the terminal auto-run result.
  * Correctly handles empty/null frontmatter status, preserves inline comments/formatting on updates, and avoids reconciliation for blocked, already-final, unknown, missing, or out-of-tree specs.
  * Verification now ignores orchestrator-generated artifact directories when checking for meaningful changes.
* **Chores**
  * Bumped release/version metadata to **0.7.8** (including marketplace/module packaging).
* **Tests**
  * Expanded coverage for the reconciliation behavior, YAML edge cases, out-of-tree skips, and updated verification gating.

### CODE REVIEW

File: src/automator/engine.py [hunk 1]
```diff
@@ -1038,6 +1038,12 @@ def _dev_phase(self, task: StoryTask) -> bool:
             advance(task, Phase.DEV_VERIFY)
             outcome = None
             if result.status == "completed":
+                # bmad-dev-auto sometimes finalizes the spec in prose (## Auto Run
+                # Result: Status done) but leaves the frontmatter status at the
+                # template default. Repair it BEFORE any frontmatter reader runs —
+                # the sync below, verify_dev, and the review-verify gate all key
+                # off the on-disk frontmatter status.
+                self._reconcile_generic_terminal_status(task, result.result_json)
```
[the remark this answers is not in the document]
Comment (pbean) [reply src/automator/engine.py:1050]: Good catch — confirmed and fixed in 61154ed.

You're right that this is real-world reachable: `bmad-auto init` only gitignores `.automator/runs|cache`, so `_bmad-output` is tracked in real projects, and the reconcile flip removes the frontmatter-status backstop that previously caught empty sessions. The old test only passed via a `_bmad-output/` gitignore hack justified by a `#2522` premise that nothing in the repo actually supports.

Fix:
- `has_changes_since` now takes an `exclude` param (mirroring `attempt_dirty`), and a new `verify.artifact_relpaths(paths)` returns the orchestrator-owned artifact folders — the same set as `Engine._protected_relpaths()`. `verify_dev` and `verify_dev_bundle` pass it, so only real source edits count as proof of work.
- `test_generic_reconcile_does_not_bypass_no_change_gate` was rewritten to drop the gitignore hack and the false comment; it now defers spec-only work on a genuinely tracked tree, while the positive reconcile path (which edits a real source file outside `_bmad-output`) still PROCEEDs.

File: src/automator/devcontract.py [hunk 2]
```diff
@@ -43,6 +43,13 @@
 DONE = "done"
 BLOCKED = "blocked"
 
+# Frontmatter statuses a half-finalized generic spec may be reconciled FROM when
+# its prose terminal `## Auto Run Result` Status is `done`. Deliberately an
+# allowlist: anything else (already-terminal done/in-review, blocked, or an
+# unknown custom token) is left untouched, so reconciliation can never override a
+# status the skill set on purpose.
+RECONCILABLE_FROM = frozenset({"", "draft", "ready-for-dev", "in-progress"})
```
[the remark this answers is not in the document]
Comment (pbean) [reply src/automator/devcontract.py:52]: Confirmed and fixed in 61154ed. Rather than drop `""` from the allowlist, we made it actionable:

- `_FM_STATUS_RE`'s value token was relaxed `[A-Za-z-]+` → `[A-Za-z-]*`, so a present-but-empty `status:` (e.g. `status:` / `status: ""`) is now matched and filled.
- `reset_spec_status` now inserts a `status:` line before the closing fence when the frontmatter has none (line-ending-aware), so a missing status is reconciled too.

Added `test_reset_status_fills_empty_value` and `test_reset_status_inserts_missing_line` to cover both. Thanks for spotting the silent no-op.

**Comment by pbean** [comment]
augment review

File: src/automator/engine.py [hunk 3]
```diff
@@ -1295,6 +1301,58 @@ def _dev_review_enabled(self) -> bool:
...
+        if not self._generic_dev():
+            return
+        spec_file = (result_json or {}).get("spec_file")
+        if not spec_file:
+            return
+        spec_path = verify.resolve_spec_path(str(spec_file), self.workspace.paths)
+        if not spec_path.is_file():
+            return
+        success_status = "in-review" if self._dev_review_enabled() else "done"
+        fm_status = verify.status_of(verify.read_frontmatter(spec_path))
```
[the remark this answers is not in the document]
Comment (pbean) [reply src/automator/engine.py]: Confirmed and fixed in 5de2aa1. You're right — a bare `status:` / `status: null` parses to YAML null, and `verify.status_of` stringifies that to `"none"` (via `str(None)`), which isn't in `RECONCILABLE_FROM`, so the most common blank form was silently skipped.

`_reconcile_generic_terminal_status` now reads the frontmatter once and normalizes a YAML-null/missing status to `""` so it matches the allowlist member (kept `status_of` untouched — it's a shared gate chokepoint and `"none"` vs `""` is immaterial to the other gates). A literal `status: none` string stays `"none"` and is still left alone.

Fixing this also surfaced a corruption bug in the round-1 fill code: a bare `status:` (no trailing space) filled to `status:done` — invalid YAML, the key drops on re-parse. `reset_spec_status` now guarantees `key: value` spacing. Added an engine test for the null-status reconcile and a devcontract test asserting the filled line re-parses cleanly. Thanks — good catch (medium was right; it had teeth).

File: src/automator/verify.py [hunk 4]
```diff
@@ -634,6 +642,26 @@ def set_frontmatter_status(path: Path, status: str) -> bool:
...
+    ``Engine._protected_relpaths``; the dev/bundle proof-of-work gate passes
+    these to ``has_changes_since`` so spec-only edits never count as real work."""
+    out: list[str] = []
+    for folder in (
+        paths.output_folder,
+        paths.implementation_artifacts,
+        paths.planning_artifacts,
+    ):
+        try:
+            out.append(folder.relative_to(paths.project).as_posix())
```
[the remark this answers is not in the document]
Comment (pbean) [reply src/automator/verify.py]: Good defensive catch — fixed in 5de2aa1. If an artifact folder is configured to equal `paths.project`, `relative_to(...).as_posix()` yields `"."`, and `:(exclude).` would exclude the whole tree, disabling the proof-of-work gate.

`artifact_relpaths` now drops empty/`"."` results. I applied the same filter to the pre-existing `Engine._protected_relpaths` (it shares the identical pattern and the same latent gap) so the two stay in sync. Added verify tests covering the `"."` guard and the artifact-only-edit exclusion.

**Comment by pbean** [comment]
augment review

File: src/automator/devcontract.py [hunk 5]
```diff
@@ -221,9 +233,22 @@ def _repl(m: re.Match[str]) -> str:
...
         changed = True
-        return f"{m.group('pre')}{m.group('q')}{new_status}{m.group('q')}{m.group('rest')}"
-
-    new_body = _FM_STATUS_RE.sub(_repl, body, count=1)
+        # Guarantee `key: value` spacing: a bare `status:` (no trailing space)
+        # would otherwise fill to `status:done` — invalid YAML, the key is lost.
+        pre = m.group("pre")
+        if not pre.endswith((" ", "\t")):
+            pre += " "
+        return f"{pre}{m.group('q')}{new_status}{m.group('q')}{m.group('rest')}"
```
[the remark this answers is not in the document]
Comment (pbean) [reply src/automator/devcontract.py]: Confirmed and fixed in 34e6927. A blank value with a trailing inline comment (`status: # tbd`, which YAML parses as null, so the round-2 normalization routes it into the fill path) abutted the value to the comment — `status: done# tbd` — and since a `#` with no preceding space isn't a YAML comment, the value became the whole `done# tbd` scalar.

`reset_spec_status._repl` now re-inserts a separating space when `rest` begins with `#`. Verified across the cases:
```
status: # intentionally blank -> status: done # intentionally blank   reparsed=done
status:   # tbd               -> status:   done # tbd                 reparsed=done
status: 'x' # keep                -> status: 'done' # keep                reparsed=done
```
Added a devcontract test asserting the comment survives and the frontmatter re-parses to `done`. Thanks!

File: src/automator/engine.py [hunk 6]
```diff
@@ -1295,6 +1305,64 @@ def _dev_review_enabled(self) -> bool:
...
+        never-regress: every deterministic verify gate still runs afterward against
+        real on-disk/git state, so this repairs bookkeeping only — it cannot pass
+        uncompleted work. Runs ahead of ``_post_dev_state_sync`` so both the story
+        (sprint) and bundle (ledger) sync, then verify, read the reconciled spec."""
+        if not self._generic_dev():
+            return
+        spec_file = (result_json or {}).get("spec_file")
+        if not spec_file:
+            return
+        spec_path = verify.resolve_spec_path(str(spec_file), self.workspace.paths)
```
[the remark this answers is not in the document]
Comment (pbean) [reply src/automator/engine.py:1330]: Good defensive catch — fixed in 34e6927. `resolve_spec_path` returns an absolute `spec_file` as-is with no containment check, and reconcile is the one path that mutates a file keyed off the session-supplied path.

Added `verify.spec_within_roots(spec_path, paths)` (mirrors the `is_relative_to` idiom already at `verify.py:265`) and gated the reconcile write on it: the spec must equal or sit under the project root or an artifact root (`output_folder` / `implementation_artifacts` / `planning_artifacts` — so a legitimately out-of-`project` artifact dir is still allowed). A spec outside those roots is left untouched and journaled as `spec-reconcile-skipped-out-of-tree`. The normal generic flow is unaffected — `find_result_artifact` globs `implementation_artifacts`, so `spec_file` is always an absolute path under that root. Added verify + engine tests for the skip path.

**Comment by pbean** [comment]
augment review

**Comment by pbean** [comment]
@coderabbitai both nitpicks addressed in d0e4d57 (test-only).

**`test_spec_within_roots`** — added the out-of-project artifact-root case: `dataclasses.replace(project, implementation_artifacts=<dir outside project>)`, then asserting a spec under that external root is still accepted. Guards against a future simplification to project-only containment, which `spec_within_roots` deliberately does not do.

**`test_has_changes_since_excludes_artifact_only_edit`** — now also writes directly under `_bmad-output` (`output_folder / "ledger.json"`) alongside the existing impl-artifact edit. The `exclude=artifact_relpaths(...)` assertion now proves the `output_folder` prefix is in the exclude set, not just the impl subdir — so root-level bundle/ledger edits can't start tripping the no-change gate unnoticed.

Full suite green (1204 passed), trunk clean. Thanks for the catches.
